### DISCUSSION TITLE
Feature Properties as Json

### DISCUSSION THREAD

**Discussion by marksmall** [visgl/deck.gl#6605] [Q&A]
I am using an `MVTLayer`, the feature's properties are complex objects, such as arrays of objects. The properties however are string representations of arrays etc. How best to convert these? I have done it in my `getFillColor`, but I need it done in multiple places, so that doesn't seem efficient. Is there a way to do it on load? I do have a couple of ideas but not sure which, if any is the most appropriate:

* Extend `MVTLayer` to override it's `getTileData` and use it.
* Use something like `onViewportLoad` or `onDataLoad`, but I've been unable to get `onDataLoad` to trigger and `onViewportLoad`, I'm not seeing the features.

Any thoughts on how best I should approach this, I must be dumb as I'd expect this to be a very common question and I've not found any example or request by anyone else on this topic.

**Comment by marksmall** [comment]
I have a solution to this, it appears to work but I'd want to test it a little longer before saying it is definitely a fix. I looked at all the callback functions and the most appropriate appears to be `dataTransform` from the `Layer` class. Here is my implementation:

```JavaScript
dataTransform: (data, previousData) =>
  data.map(feature => ({
    ...feature,
    properties: Object.entries(feature.properties).reduce(
      (prev, [key, value]) => {
        let result = null;
        try {
          result = JSON.parse(value);
        } catch (error) {}
        return {
          ...prev,
          [key]: result ?? value,
        };
      },
    ),
  }))
```

I'm not saying this is the best answer, but from my testing of 5 of our datasets, it is just as performant as our current solution.

**Reply by arthurgailes** [reply]
I'd start with `onTileLoad`, then you'd have something like:

```
onTileLoad(tile){
  if(!tile.content) return;
  tile.content.points.properties.forEach(prop=>{ // you might have polygons or lines
    Object.assign(prop, {newArray: yourFunction(prop.key)})
  })
}
```
That will assign the prop for the lifecycle of the tile, so you're not repeating the operation.

https://deck.gl/docs/api-reference/geo-layers/tile-layer#tile

**Reply by marksmall** [reply]
Thanks for that, I don't see the `dataTransform` repeating, I see for each `onTileLoad` a corresponding `dataTransform`, but I'd like follow the common wisdom on this though, so will try it out with your suggested code, thanks.

**Reply by arthurgailes** [reply]
:) Can't claim any particular wisdom, it's just how I've been doing it.
